### PR TITLE
feat: add ipv4/ipv6 dual stack support

### LINKED ISSUE #184: Add IPv4/IPv6 Dual Stack Support
Currently, EG only supports IPv4. For example, an Infra IR proxy listen address represents an IPv4 address. The code should be audited for IPv4-ism's and updated to support a dual-stack IPv4/IPv6 environment properly.

xref: https://github.com/envoyproxy/gateway/pull/183#issuecomment-1201507246

https://kubernetes.io/docs/concepts/services-networking/dual-stack/

### PR DESCRIPTION
**Pull request by juwon8891**
**What type of PR is this?**

feat: add ipv4/ipv6 dual stack support

**What this PR does / why we need it**:
This PR adds IPv4/IPv6 dual stack support to Envoy Gateway. It implements test cases for routing to IPv6-only, dual-stack, and IPv4-only services using HTTPRoute resources. This enhancement is crucial for supporting modern network environments and ensuring compatibility with both IPv4 and IPv6 infrastructures.

Key changes include:
1. New test file `backend_dualstack.go` for testing IPv6 and dual-stack backends
2. New test file `httproute_dualstack.go` for testing HTTPRoute support with various IP versions
3. New manifest files for defining resources used in dual-stack tests
4. Implementation of `getDnsLookupFamily` function to control DNS lookup behavior

**Which issue(s) this PR fixes**:

Fixes #184 

Additional Notes:
- Tests are skipped in IPv4-only environments (`IP_FAMILY=ipv4`)
- All tests run in IPv6 or dual-stack environments (`IP_FAMILY=ipv6` or `IP_FAMILY=dual`)

### CODE REVIEW

File: api/v1alpha1/backend_types.go [hunk 1]
```diff
@@ -73,10 +73,11 @@ type BackendEndpoint struct {
 // https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-socketaddress
 type IPEndpoint struct {
 	// Address defines the IP address of the backend endpoint.
+	// Supports both IPv4 and IPv6 addresses.
 	//
-	// +kubebuilder:validation:MinLength=7
-	// +kubebuilder:validation:MaxLength=15
-	// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$`
+	// +kubebuilder:validation:MinLength=1
```
Comment (arkodg) [inline api/v1alpha1/backend_types.go:78]: curious why is this 1 ? guessing `::1` is the smallest which is 3 chars

File: test/cel-validation/envoyproxy_test.go [hunk 2]
```diff
@@ -949,35 +949,6 @@ func TestEnvoyProxyProvider(t *testing.T) {
 			},
 			wantErrors: []string{},
 		},
-		{
```
Comment (arkodg) [inline test/cel-validation/envoyproxy_test.go:952]: why is this removed ?

File: test/e2e/testdata/backend-dualstack.yaml [hunk 3]
```diff
@@ -0,0 +1,95 @@
+apiVersion: v1
+kind: Service
```
Comment (arkodg) [inline test/e2e/testdata/backend-dualstack.yaml:2]: should this be removed since its being tested in `httproute-dualstack.yaml` ?

**Comment by arkodg** [comment]
hey @juwon8891 we probably also need to add some logic in https://github.com/envoyproxy/gateway/blob/4b8c2f56dc45534c42b6513b126082f49a009658/internal/gatewayapi/route.go#L1328 which selects specific endpoints based on the [ipFamilyPolicy field](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceSpec)

File: api/v1alpha1/backend_types.go [hunk 4]
```diff
@@ -73,10 +73,11 @@ type BackendEndpoint struct {
...
+	// +kubebuilder:validation:MaxLength=45
+	// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$|^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$`
```
Comment (zirain) [inline api/v1alpha1/backend_types.go:80]: should we raise this up in upstream, request for an easy way to validate ip(for both ipv4 and ipv6)?

File: site/content/en/v1.1/api/extension_types.md
Comment (zirain) [inline site/content/en/v1.1/api/extension_types.md:1]: why change `v1.1`?

File: test/e2e/testdata/backend-dualstack.yaml (hunk 3 above)
Comment (juwon8891) [reply test/e2e/testdata/backend-dualstack.yaml:2]: Yes I think it can be remove.

File: test/cel-validation/envoyproxy_test.go (hunk 2 above)
Comment (juwon8891) [reply test/cel-validation/envoyproxy_test.go:952]: Sorry, I did test, will revert.

File: api/v1alpha1/shared_types.go [hunk 5]
```diff
@@ -297,7 +297,7 @@ type KubernetesServiceSpec struct {
 	// This field has been deprecated in Kubernetes, but it is still used for setting the IP Address in some cloud
 	// providers such as GCP.
 	//
-	// +kubebuilder:validation:XValidation:message="loadBalancerIP must be a valid IPv4 address",rule="self.matches(r\"^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$\")"
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.matches('^([0-9]{1,3}\\\\.[0-9]{1,3}\\\\.[0-9]{1,3}\\\\.[0-9]{1,3}|[a-fA-F0-9:]+)$')",message="loadBalancerIP must be a valid IPv4 or IPv6 address"
```
[the remark this answers is not in the document]
Comment (juwon8891) [reply api/v1alpha1/shared_types.go:300]: Thanks I will revert.

File: api/v1alpha1/backend_types.go (hunk 4 above)
Comment (juwon8891) [reply api/v1alpha1/backend_types.go:80]: I think so, it need to easy way to validate ip.

File: site/content/en/v1.1/api/extension_types.md
Comment (juwon8891) [reply site/content/en/v1.1/api/extension_types.md:1]: Sorry, I did test, will revert.

File: test/e2e/tests/httproute_dualstack.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		if ipFamily == "ipv4" {
+			t.Skip("Skipping IP Version Routing test as IP_FAMILY is set to 'ipv4'")
+			return
+		}
+
+		ns := "gateway-conformance-infra"
+		gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
+
+		t.Run("HTTPRoute to IPv6 only service", func(t *testing.T) {
+			routeNN := types.NamespacedName{Name: "ipv6-only-route", Namespace: ns}
```
Comment (arkodg) [inline test/e2e/tests/httproute_dualstack.go:40]: can we also check for the IP type in the `X-Forwarded-For` response header ?

**Comment by arkodg** [comment]
hey @juwon8891 lint is failing
```

internal/xds/translator/testdata/in/xds-ir/tcp-listener-ipfamily.yaml
  Error: 7:5 [indentation] wrong indentation: expected 6 but found 4
  Error: 11:9 [indentation] wrong indentation: expected 10 but found 8
  Error: 12:11 [indentation] wrong indentation: expected 12 but found 10
  Error: 13:32 [new-line-at-end-of-file] no new line character at the end of file
  ```

File: api/v1alpha1/envoyproxy_types.go [hunk 7]
```diff
@@ -134,6 +134,13 @@ type EnvoyProxySpec struct {
 	// These settings are applied on backends for which TLS policies are specified.
 	// +optional
 	BackendTLS *BackendTLSConfig `json:"backendTLS,omitempty"`
+
+	// IPFamily specifies the IP family for the EnvoyProxy.
+	// It can be ipv4, ipv6, or dual.
+	// +kubebuilder:default=ipv4
```
Comment (zirain) [inline api/v1alpha1/envoyproxy_types.go:140]:
```suggestion
	// +kubebuilder:default= IPv4
```

File: api/v1alpha1/envoyproxy_types.go [hunk 8]
```diff
@@ -134,6 +134,13 @@ type EnvoyProxySpec struct {
...
+	// +kubebuilder:validation:Enum=ipv4;ipv6;dual
```
Comment (zirain) [inline api/v1alpha1/envoyproxy_types.go:141]:
```suggestion
	// +kubebuilder:validation:Enum= IPv4;IPv6;DualStack
```

Comment (zirain) [reply api/v1alpha1/envoyproxy_types.go:141]: maybe you can seperated this into another PR, and add some cel tests.

File: api/v1alpha1/envoyproxy_types.go [hunk 9]
```diff
@@ -410,6 +417,18 @@ type EnvoyProxyList struct {
...
+// IPFamily defines the IP family to use for the Envoy proxy.
+type IPFamily string
+
+const (
+	// IPv4 defines the IPv4 family.
+	IPv4 IPFamily = "IPv4"
+	// IPv6 defines the IPv6 family.
+	IPv6 IPFamily = "IPv6"
+	// DualStack defines the dual-stack family.
+	DualStack IPFamily = "DualStack"
```
Comment (zhaohuabing) [inline api/v1alpha1/envoyproxy_types.go:429]: Nit: Could we add a line to explain how "DualStack" impacts the Envoy proxy fleet here? For example, Envoy proxy will listen on both the IPV4 and IPV6 addresses for incoming traffic.

File: internal/ir/xds.go [hunk 10]
```diff
@@ -223,8 +223,20 @@ type CoreListenerDetails struct {
 	ExtensionRefs []*UnstructuredRef `json:"extensionRefs,omitempty" yaml:"extensionRefs,omitempty"`
 	// Metadata is used to enrich envoy resource metadata with user and provider-specific information
 	Metadata *ResourceMetadata `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+	// IPFamily specifies the IP address family for the gateway.
+	// It can be IPv4, IPv6, or Dual.
+	IPFamily IPFamily `json:"ipFamily,omitempty" yaml:"ipFamily,omitempty"`
 }
 
+// IPFamily represents the IP address family for the gattype IPFamily {eway.
```
Comment (zhaohuabing) [inline internal/ir/xds.go:231]:
```suggestion
// IPFamily specifies the IP address family used by the Gateway for its listening ports.

File: api/v1alpha1/backend_types.go (hunk 4 above)
Comment (zirain) [reply api/v1alpha1/backend_types.go]: https://github.com/kubernetes/kubernetes/blob/36122d5a9bbb87f484149e3eb11bb4208228c36f/staging/src/k8s.io/apiserver/pkg/cel/library/ip.go#L102

can we use this?

File: api/v1alpha1/envoyproxy_types.go [hunk 11]
```diff
@@ -134,6 +134,13 @@ type EnvoyProxySpec struct {
 	// These settings are applied on backends for which TLS policies are specified.
 	// +optional
 	BackendTLS *BackendTLSConfig `json:"backendTLS,omitempty"`
+
+	// IPFamily specifies the IP family for the EnvoyProxy.
+	// It can be ipv4, ipv6, or dual.
+	// +kubebuilder:default=ipv4
+	// +kubebuilder:validation:Enum=ipv4;ipv6;dual
+	// +optional
+	IPFamily *IPFamily `json:"ipFamily,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/envoyproxy_types.go:143]: This only affects the Gateway listener ports, right?

Comment (zhaohuabing) [reply api/v1alpha1/envoyproxy_types.go:143]: Could we mention the default behavior here if IPFamily is not specified?

**Review by zhaohuabing — COMMENTED** [review]
@juwon8891 It's fantastic to see this much-anticipated feature implemented! Greate job! I have a few minor comments and a question.

Please correct me if I'm mistaken, this PR does two things:
* Adds support for IPV6 at the Backend
* Adds IPV6/dual stack support for the Gateway listening ports

The backend clusters won't support "dual-stack". EG only uses the first available address of the Pod/Service/Backend, wich is either an IPV4 or IPV6 address, but not both simultaneously.

File: api/v1alpha1/shared_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply api/v1alpha1/shared_types.go]: Could it be an IPV6 address if it's a DualStack/IPV6-only cluster?

File: internal/xds/translator/cluster.go [hunk 12]
```diff
@@ -83,7 +84,7 @@ func buildEndpointType(settings []*ir.DestinationSetting) EndpointType {
 func buildXdsCluster(args *xdsClusterArgs) *clusterv3.Cluster {
 	cluster := &clusterv3.Cluster{
 		Name:            args.name,
-		DnsLookupFamily: clusterv3.Cluster_V4_ONLY,
+		DnsLookupFamily: getDNSLookupFamily(args.settings),
```
Comment (zhaohuabing) [inline internal/xds/translator/cluster.go:87]: Should we use "clusterv3.Cluster_ALL"?

File: api/v1alpha1/envoyproxy_types.go [hunk 13]
```diff
@@ -134,6 +134,13 @@ type EnvoyProxySpec struct {
 	// These settings are applied on backends for which TLS policies are specified.
 	// +optional
 	BackendTLS *BackendTLSConfig `json:"backendTLS,omitempty"`
+
+	// IPFamily specifies the IP family for the EnvoyProxy.
+	// It can be ipv4, ipv6, or dual.
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_types.go:139]:
```suggestion
	// It can be Ipv4, Ipv6, or Dual.
```

File: api/v1alpha1/envoyproxy_types.go [hunk 14]
```diff
@@ -134,6 +134,13 @@ type EnvoyProxySpec struct {
 	// These settings are applied on backends for which TLS policies are specified.
 	// +optional
 	BackendTLS *BackendTLSConfig `json:"backendTLS,omitempty"`
+
+	// IPFamily specifies the IP family for the EnvoyProxy.
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_types.go:138]:
```suggestion
	// IPFamily specifies the IP family for the EnvoyProxy fleet.
```

File: internal/ir/xds.go [hunk 15]
```diff
@@ -223,8 +223,20 @@ type CoreListenerDetails struct {
...
+type IPFamily string
+
+const (
+	IPv4 IPFamily = "ipv4"
```
Comment (arkodg) [inline internal/ir/xds.go:235]: prefer same naming as API, so `Ipv4`, `Ipv6` ...

File: internal/ir/xds.go [hunk 16]
```diff
@@ -223,8 +223,20 @@ type CoreListenerDetails struct {
 	ExtensionRefs []*UnstructuredRef `json:"extensionRefs,omitempty" yaml:"extensionRefs,omitempty"`
 	// Metadata is used to enrich envoy resource metadata with user and provider-specific information
 	Metadata *ResourceMetadata `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+	// IPFamily specifies the IP address family for the gateway.
+	// It can be IPv4, IPv6, or Dual.
+	IPFamily IPFamily `json:"ipFamily,omitempty" yaml:"ipFamily,omitempty"`
```
Comment (arkodg) [inline internal/ir/xds.go:228]: my preference would be to not use IPFamily in this layer but convert the `Address` field into a list and pass `::` it and in the xds layer iterate deal with this list and convert it into `Address` and `AdditionalAddresses` fields

File: internal/xds/translator/listener.go [hunk 17]
```diff
@@ -170,6 +170,28 @@ func buildXdsTCPListener(name, address string, port uint32, keepalive *ir.TCPKee
 			},
 		},
 	}
+
+	if ipFamily == ir.Dual {
```
Comment (arkodg) [inline internal/xds/translator/listener.go:174]: there are 3 options here - ipv4 only, ipv6 only and dual, missing ipv6

**Comment by arkodg** [comment]
hey @juwon8891 looks like some gateway api tests got wiped out, can you add them back ?

File: test/e2e/tests/backend_dualstack.go [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+func init() {
+	ipFamily := os.Getenv("IP_FAMILY")
+	if ipFamily == "" || ipFamily == "ipv4" {
+		BackendDualStackTest = suite.ConformanceTest{
+			ShortName: "BackendDualStack",
+			Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+				t.Skip("Skipping BackendDualStack test as IP_FAMILY is not set or set to ipv4")
+			},
+		}
+	}
```
Comment (zirain) [inline test/e2e/tests/backend_dualstack.go:30]: this can be moved into `BackendDualStackTest` like, instead of in `init`.
```
if ipFamily == "" || ipFamily == "ipv4"{
    t.Skip("Skipping BackendDualStack test as IP_FAMILY is not set or set to ipv4")
}
```

File: test/e2e/tests/backend_dualstack.go [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+var BackendDualStackTest = suite.ConformanceTest{
+	ShortName:   "BackendDualStack",
+	Description: "Test IPv6 and Dual Stack support for backends",
+	Manifests:   []string{"testdata/backend-dualstack.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+		gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
+		ipFamily := os.Getenv("IP_FAMILY")
+		if ipFamily == "ipv6" || ipFamily == "dual" {
```
Comment (zirain) [inline test/e2e/tests/backend_dualstack.go:42]: can we simply this with `switch-case`?

File: test/e2e/tests/httproute_dualstack.go [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+	ipFamily := os.Getenv("IP_FAMILY")
+	if ipFamily == "" {
+		HTTPRouteDualStackTest = suite.ConformanceTest{
+			ShortName: "HTTPRouteDualStack",
+			Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+				t.Skip("Skipping IP Version Routing test as IP_FAMILY is not set")
+			},
+		}
+	}
+	ConformanceTests = append(ConformanceTests, HTTPRouteDualStackTest)
```
Comment (zirain) [inline test/e2e/tests/httproute_dualstack.go:31]: same as `backend_dualstack.go`

File: test/e2e/tests/httproute_dualstack.go [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+var HTTPRouteDualStackTest = suite.ConformanceTest{
+	ShortName:   "HTTPRouteDualStack",
+	Description: "Test HTTPRoute support for IPv6 only, dual-stack, and IPv4 only services",
+	Manifests:   []string{"testdata/httproute-dualstack.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+		gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
+		ipFamily := os.Getenv("IP_FAMILY")
+
+		if ipFamily == "ipv6" || ipFamily == "dual" {
```
Comment (zirain) [inline test/e2e/tests/httproute_dualstack.go:43]: same as `backend_dualstack.go`

File: api/v1alpha1/envoyproxy_types.go [hunk 22]
```diff
@@ -134,6 +134,23 @@ type EnvoyProxySpec struct {
...
+	//
+	// If not specified, the behavior depends on the environment:
+	// - In IPv4-only environments, it defaults to IPv4.
+	// - In dual-stack environments, it will use both IPv4 and IPv6 (equivalent to DualStack).
+	// - In IPv6-only environments, it defaults to IPv6.
+	//
+	// It's recommended to explicitly set this field to ensure consistent behavior
+	// across different environments.
+	//
+	// +kubebuilder:default=IPv4
```
Comment (zirain) [inline api/v1alpha1/envoyproxy_types.go:150]: > - In IPv4-only environments, it defaults to IPv4.
> - In IPv6-only environments, it defaults to IPv6.

so I think you should remove the default?

**Comment by zhaohuabing** [comment]
@juwon8891 Do you have time to address the comments this week? We want to ensure dual stack support is ready to be included in the coming v1.3.

**Comment by juwon8891-sidewalk** [comment]
Yes, I'm working on it. I'll answer as soon as possible

**Comment by juwon8891** [comment]
Yes, I'm working on it. I'll answer as soon as possible

File: internal/xds/translator/cluster.go (hunk 12 above)
Comment (juwon8891) [reply internal/xds/translator/cluster.go:87]: Can we adopt an idea like this? This is the dns selection method of isio.
image
image

Comment (zhaohuabing) [reply internal/xds/translator/cluster.go:87]: @juwon8891 We can use the `EnvoyProxy.IPFamily` to determine the `DnsLookupFamily`.

File: internal/xds/translator/cluster.go [hunk 23]
```diff
@@ -81,9 +84,13 @@ func buildEndpointType(settings []*ir.DestinationSetting) EndpointType {
 }
 
 func buildXdsCluster(args *xdsClusterArgs) *clusterv3.Cluster {
+	dnsLookupFamily := clusterv3.Cluster_V4_ONLY
```
Comment (zhaohuabing) [inline internal/xds/translator/cluster.go:87]: What if "*args.ipFamily == egv1a1.IPv6"? Do we want to support IPv6 single stack? @arkodg @zirain

Comment (arkodg) [reply internal/xds/translator/cluster.go:87]: this field is tied to the cluster/endpoint config for control plane communication b/w envoy proxy and envoy gateway, its safe to make this `clusterv3.Cluster_ALL` and rely on the DNS service to provide the right info

File: internal/xds/translator/cluster.go (hunk 12 above)
Comment (arkodg) [reply internal/xds/translator/cluster.go]: @zhaohuabing `EnvoyProxy.IPFamily` is meant to configure Envoy listener and Envoy Proxy service config, not backend endpoint config, which is determined based on info in the backendRef `Service` and its corresponding `EndpointSlice` resources

Comment (arkodg) [reply internal/xds/translator/cluster.go]: the API can be made clearer to avoid this confusion

File: test/e2e/tests/httproute_dualstack.go [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+		Request: http.Request{
+			Path: path,
+		},
+		Response: http.Response{
+			StatusCode: 200,
+		},
+		Namespace: ns,
+	}
+
+	http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expectedResponse)
```
Comment (zirain) [inline test/e2e/tests/httproute_dualstack.go:73]: https://github.com/envoyproxy/gateway/blob/4b8c2f56dc45534c42b6513b126082f49a009658/test/e2e/tests/utils.go#L535

can be used to verify pod ip/port

File: internal/xds/translator/cluster.go (hunk 23 above)
Comment (arkodg) [reply internal/xds/translator/cluster.go:87]: @juwon8891 can we update this to `clusterv3.Cluster_ALL` by default ?

File: internal/xds/translator/listener.go [hunk 25]
```diff
@@ -199,8 +230,6 @@ func buildXdsQuicListener(name, address string, port uint32, accesslog *ir.Acces
 			DownstreamSocketConfig: &corev3.UdpSocketConfig{},
 			QuicOptions:            &listenerv3.QuicProtocolOptions{},
 		},
-		// Remove /healthcheck/fail from endpoints that trigger a drain of listeners for better control
```
Comment (arkodg) [inline internal/xds/translator/listener.go:202]: does this need to be reverted ?

**Comment by arkodg** [comment]
hey @juwon8891, added some minor comments, but overall LGTM !

File: internal/gatewayapi/backendtlspolicy.go [hunk 26]
```diff
@@ -157,6 +157,20 @@ func backendTLSTargetMatched(policy gwapiv1a3.BackendTLSPolicy, target gwapiv1a2
 	return false
 }
 
+func getIPFamily(envoyProxy *egv1a1.EnvoyProxy) ir.IPFamily {
```
[the remark this answers is not in the document]
Comment (juwon8891) [reply internal/gatewayapi/backendtlspolicy.go:160]: ok It will be move.

File: internal/xds/translator/listener.go (hunk 25 above)
Comment (juwon8891) [reply internal/xds/translator/listener.go:202]: Yes it need to revert.

File: test/e2e/tests/httproute_dualstack.go (hunk 24 above)
Comment (juwon8891) [reply test/e2e/tests/httproute_dualstack.go:73]: I'm still working on this test part. I'll post it separately as another PR.

File: internal/xds/translator/cluster.go (hunk 12 above)
Comment (zhaohuabing) [reply internal/xds/translator/cluster.go]: Oh, I see. I got the EnvoyProxy listener IP family settings with the Backend IP families.

**Comment by zirain** [comment]
@juwon8891 please run `make gen-check` to pass the `gen-check` CI.

File: internal/ir/xds.go [hunk 27]
```diff
@@ -224,8 +224,20 @@ type CoreListenerDetails struct {
...
```
Comment (zirain) [inline internal/ir/xds.go:229]: can you make it a pointer? so that this PR won't be that huge.

**Comment by arkodg** [comment]
@juwon8891 DCO is failing, can you sign your commits

File: internal/ir/xds.go (hunk 27 above)
Comment (juwon8891-sidewalk) [reply internal/ir/xds.go]: > I'll post it separately as another PR.

I changed here.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by guydc — APPROVED** [review]
LGTM!
